includes/subscriptions/class-wc-payments-product-service.php

```diff
@@ -478,10 +478,78 @@ public function unarchive_price( string $wcpay_price_id, $test_mode = null ) {
 		$this->payments_api_client->update_price( $wcpay_price_id, $data );
 	}
 
+	/**
+	 * Prevents the subscription interval to be greater than 1 for yearly subscriptions.
+	 *
+	 * @param int $product_id Post ID of the product.
+	 */
+	public function limit_yearly_subscription_intervals_on_save_product( $product_id ) {
+		// Skip products that aren't subscriptions.
+		$product = wc_get_product( $product_id );
+		if (
+			! $product ||
+			! WC_Subscriptions_Product::is_subscription( $product ) ||
+			empty( $_POST['_wcsnonce'] ) ||
+			! wp_verify_nonce( sanitize_key( $_POST['_wcsnonce'] ), 'wcs_subscription_meta' )
+		) {
+			return;
+		}
+
+		if (
+			isset( $_REQUEST['_subscription_period'] ) &&
+			'year' === $_REQUEST['_subscription_period'] &&
+			isset( $_REQUEST['_subscription_period_interval'] ) &&
+			'1' !== $_REQUEST['_subscription_period_interval']
+		) {
+			// Prevent WC Subs Core from saving the interval when it's invalid.
+			$_REQUEST['_subscription_period_interval'] = '1';
+
+			// Add an admin notice to let the merchant know about this limitation.
+			wcs_add_admin_notice( __( "The period interval of a subscription can't be greater than a year.", 'woocommerce-payments' ), 'error' );
+		}
+	}
+
+	/**
+	 * Prevents the subscription interval to be greater than 1 for yearly subscription variations.
+	 *
+	 * @param int $product_id Post ID of the variation.
+	 * @param int $index Variation index in the incoming array.
+	 */
+	public function limit_yearly_subscription_intervals_on_save_product_variation( $product_id, $index ) {
+		// Skip products that aren't subscriptions.
+		$product = wc_get_product( $product_id );
+		if (
+			! $product ||
+			! WC_Subscriptions_Product::is_subscription( $product ) ||
+			empty( $_POST['_wcsnonce_save_variations'] ) ||
+			! wp_verify_nonce( sanitize_key( $_POST['_wcsnonce_save_variations'] ), 'wcs_subscription_variations' )
+		) {
+			return;
+		}
+
+		if (
+			isset( $_POST['variable_subscription_period'][ $index ] ) &&
+			'year' === $_POST['variable_subscription_period'][ $index ] &&
+			isset( $_POST['variable_subscription_period_interval'][ $index ] ) &&
+			'1' !== $_POST['variable_subscription_period_interval'][ $index ]
+		) {
```

Comment:
Btw, completely unrelated to this PR but a quick question because I just noticed it. I'm aware those docs are for the WC Subscriptions extension, but is that snippet intended to work for WCPay's subs as well? Asking because it looks like the class`WC_Subscriptions` doesn't exist there. Wondering if we should update the snippet's `WC_Subscriptions::append_numeral_suffix()` to `wcs_append_numeral_suffix()`, like in the snippet in the "How can I offer a longer subscription length?" section below.

Reply:
> I'm aware those docs are for the WC Subscriptions extension, but is that snippet intended to work for WCPay's subs as well?

Wow, good eye!
`wcs_append_numeral_suffix()` is a new function that is getting introduced in Subscriptions 4.0 (it's come early with WCPay Subscriptions) but yes, you're spot on! We'll have to update this code snippet to not use a deprecated function but also so that it works with WCPay Subscriptions feature as well 💯